### DISCUSSION TITLE
[LEMFO LEMP Full Android 9.1] Unable to start Bluetooth Service (Dexcom / MiaoMiao / Bubble)

### DISCUSSION THREAD

**Discussion by parapenT1sta** [Navid200/xDrip#161] [Testing]
A friend of mine bought a LEMFO LEMP with Android 9.1 hoping was going to work like the previous models running Android 7.1.1 with MiaoMiao and Bubble but in fact he is unable to use it.
xDrip cannot start the Bluetooth Service, at least it seems. Since I have Dexcom, I tried to connect Dexcom on xDrip and the same issue, tried all settings, then I installed Dexcom Built Your Own App and it connects with no issue.

xDrip with Dexcom: Phone Service State: Service Stopped
xDrip with Bubble or MiaoMiao: Phone Service State: Full Service Stopped

Is anything that I can debug for further investigation or this is a case where nothing can be done? 

![PXL_20211222_133826035](https://user-images.githubusercontent.com/608650/147102779-961e657b-9ce6-4d9e-a91c-1c7c4593d19c.jpg)
![PXL_20211222_133001731](https://user-images.githubusercontent.com/608650/147102784-154e6b2d-1703-4b59-9af8-de87e095e7d6.jpg)
![PXL_20211222_132054744](https://user-images.githubusercontent.com/608650/147102786-a33e572c-86d7-4f47-b9fc-8e59b586e053.jpg)

UPDATE: https://github.com/Navid200/xDrip/discussions/161#discussioncomment-1865793

**Comment by Navid200** [comment]
The issues platform has become cluttered with items that should really be in a forum for discussing many different subjects.
Unfortunately, the issues platform has become the first place many go to.
One of the unfortunate consequences is that there are now many closed issues that no one looks at.  Even though a feature request or bug that has been resolved, and is now a closed issue, still contains valuable information. 

We have added guidelines for using the discussions, when someone attempts to open an issue.
You must have seen those.  The idea is to encourage everyone to go to the discussions first instead of the issues.

You have now added a link to the closed issue.
Would you please delete that link?
If this discussion results in a need for an issue (bug or feature request) to be opened, we will then ask you to include a link in the issue after reopening it.  But, for now, please help us encourage everyone to come to the discussions where they can find out everything needed about what is going on.
Thanks for your cooperation in advance.

In the meantime, I will ask the right person to have a look at this discussion thread.

**Reply by parapenT1sta** [reply]
Link deleted...

**Reply by Navid200** [reply]
Thanks

**Comment by tolot27** [comment]
@parapenT1sta You could enable "Store logs for troubleshooting" in extra logging. It requires enabling the engineering mode and maybe a restart of xDrip. It creates a "xdriplogcat.txt" on the watch. Hopefully, something is logged. You can also enable verbose logging for the various Dexcom and Bluetooth related classes.

BTW: Looks like the transmitter is an old one with reused batteries.

**Reply by parapenT1sta** [reply]
Can you please check this information: https://github.com/Navid200/xDrip/discussions/161#discussioncomment-1865793?

Btw, I couldn't find the xdriplogcat.txt on the watch. The option to save is selected but no file is created I think also because this Full Android 9.1 watch is detected as an Android Wear.

**Comment by Navid200** [comment]
If you install xDrip again, does it still fail to connect to the Dexcom transmitter?
If you then, uninstall xDrip and install the build your own Dexcom app, does it connect again?
If you go back and forth between xDrip and the BYO Dexcom app, does it always work with the Dexcom app and does it always fail with xDrip?

**Reply by parapenT1sta** [reply]
@Navid200 @tolot27 I tried your steps already but I think I found the issue running the extra logging Ob1G5CollectionService:v, just let me know if it helps, this is the only information I could get. It seems this Full Android 9.1 Watch is detected as an Android Wear (this option is disabled on xDrip) or is not related? "We are running on Android Wear"

Can you provide me the extra logging tag for Libre Bluetooth service, is it DexCollectionService?

![PXL_20211224_000513351_2](https://user-images.githubusercontent.com/608650/147301511-4895844f-7c7b-47df-9779-a078b966642f.jpg)

![PXL_20211224_000520661 MP](https://user-images.githubusercontent.com/608650/147301543-835284ba-4ad2-4fd2-919a-e982e249c451.jpg)

![PXL_20211224_000534290](https://user-images.githubusercontent.com/608650/147301521-aedd7059-3ca9-4e57-9540-800fc2b2cfbf.jpg)

**Comment by parapenT1sta** [comment]
@Navid200 and @tolot27 Is it possible to get a version for testing proposes setting the device as "Phone" all the time just to avoid Android Wear detection and to verify if the issue is solved for Dexcom and MiaoMiao/Bubble on this device.

As mentioned on the previous post this is the order of the logs:
```
Resetting internal state by request
Not registering pairing receiver on Device Android 8+
Checking model: LEMP
We are running on Android Wear
WAKE UP WAKE UP WAKE UP WAKE UP
Status: Wake Up
Stopping Service due to shouldServiceRun() result
Status: Stopping
Status: Shutting down
DEBUG: killing stop scan task
Got exception unregistering pairing receiver: java.lang.IllegalArgumentException: Receiver not register: com.eveningoutpost.dexdrip.services.Ob1G5CollectionService$2@652d9e7
Status: Service Stopped
```

**Reply by Navid200** [reply]
This is a full Android watch, right?
Have you disabled Android wear on it?

**Comment by parapenT1sta** [comment]
@Navid200 Yes, as I mentioned on the previous comment, this is a Full Android Watch (works like a phone) and Android Wear option is disable.

**Reply by Navid200** [reply]
Android wear must be enabled for that log to show up:
https://github.com/NightscoutFoundation/xDrip/blob/5bf48f918ce10bd59c8eb855551c2ae3d663ded7/app/src/main/java/com/eveningoutpost/dexdrip/Services/Ob1G5CollectionService.java#L973-L974

**Comment by parapenT1sta** [comment]
@Navid200 You're getting annoying! If I'm telling you that I have the option is disabled is because is DISABLED!
Take my settings file saved to check...
`<boolean name="wear_sync" value="false" />`
[com.eveningoutpost.dexdrip_preferences.zip](https://github.com/Navid200/xDrip/files/7776083/com.eveningoutpost.dexdrip_preferences.zip)

**Reply by Navid200** [reply]
I don't think I can help you.
Hopefully, someone else can

**Reply by parapenT1sta** [reply]
Maybe something is wrong on the code assuming that the full watch is an Android Wear and skipping the option. 

But at least did you check my preferences? Because you were so right that I have the option enabled and this is annoying for me as I am here to try to provide all right the information to check if Xdrip can Work at least with Dexcom as per official Dexcom app works perfectly fine.

Try to accept the information people are giving to you...

**Comment by parapenT1sta** [comment]
**Finally I made it!** Modified xDrip code with some workarounds and built my own app and now Dexcom G5 it's working with LEMFO LEMP Full Android 9.1.  Next step is to modify DexCollectionService to make it work with MiaoMiao and Bubble. If anyone is facing the same issue I can sent my xDrip version.

![G5](https://user-images.githubusercontent.com/608650/147416144-18dc5ab6-4716-4c8b-a22d-1e6c80158555.jpg)

**Comment by parapenT1sta** [comment]
Now, also working with Bubble! Since I'm not a programmer at all and just used workarounds in the code, I don't know how to fix it properly but at least it's working...

![PXL_20211226_183857832 NIGHT](https://user-images.githubusercontent.com/608650/147417287-4a0bb066-b30f-427f-b25b-74c4753c3f16.jpg)

**Comment by parapenT1sta** [comment]
Changes that I've done on **Ob1G5CollectionService.java** to make it work with G5. I will describe my thoughts on the code changes as Bluetooth Service was being shutdown due wrong Android wear detection. 
Like I mentioned before, I'm not a programmer and this was just a try, very amateur. Used **//**  to make the code as comments.

960 - Looks like even being an Android 9.1 watch needs this command to make it work: _**registerReceiver(mPairingRequestRecevier, pairingRequestFilter);**_ , so I increased the SDK version from 26 to 29.
`                if (Build.VERSION.SDK_INT < 29) {`

974-975 - The watch is being wrongly detected as Android Wear, so I hid the command _**wear_broadcast = Pref.getBooleanDefaultFalse("ob1_wear_broadcast");**_
```
 //                   UserError.Log.d(TAG, "We are running on Android Wear");
 //                   wear_broadcast = Pref.getBooleanDefaultFalse("ob1_wear_broadcast");
```

1011-1016 - Avoid to stop the service
```
 //           if (!shouldServiceRun()) {
 //               UserError.Log.d(TAG, "Stopping service due to shouldServiceRun() result");
 //               msg("Stopping");
 //               stopSelf();
 //               return START_NOT_STICKY;
 //           }
```

**Reply by tolot27** [reply]
> ```
>  //                   UserError.Log.d(TAG, "We are running on Android Wear");
>  //                   wear_broadcast = Pref.getBooleanDefaultFalse("ob1_wear_broadcast");
> ```

Before that lines there is a check:
`android_wear = (getResources().getConfiguration().uiMode & Configuration.UI_MODE_TYPE_MASK) == Configuration.UI_MODE_TYPE_WATCH;`
Please can you change it to:
`android_wear = getPackageManager().hasSystemFeature(PackageManager.FEATURE_WATCH);`

and test whether it still reports "We are running on Android Wear" or not?

**Reply by parapenT1sta** [reply]
@tolot27  I cannot compile it, I get the following error:  Cannot Find Symbol Variable PackManager :972

`                android_wear = getPackageManager().hasSystemFeature(PackageManager.FEATURE_WATCH);
                                                                    ^
symbol:   variable PackageManager
location: class Ob1G5CollectionService`

**Reply by tolot27** [reply]
Just `import android.content.pm.PackageManager;`

**Reply by parapenT1sta** [reply]
@tolot27 WOW! It's working with your changes!! 👍 At least for Dexcom I feel the issue is now fixed in the proper way. 

Can you help me changing the code for DexCollectionService? To fix the issue with MiaoMiao and Bubble (Libre Bluetooth)
The approach in the code it's different in my perspective, I could make it working but with a lot hidden code.

**Reply by parapenT1sta** [reply]
@tolot27 In https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java I basically hid lines from 1230 to 1267 to avoid service to be Stopped.

Then I can check from line 821 to 829 some reference to Wear.
`    @SuppressLint("ObsoleteSdkInt")
    private static boolean shouldServiceRun() {
        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) return false;
        final boolean result = (DexCollectionType.hasXbridgeWixel() || DexCollectionType.hasBtWixel())
                && ((!Home.get_forced_wear() && (((UiModeManager) xdrip.getAppContext().getSystemService(UI_MODE_SERVICE)).getCurrentModeType() != Configuration.UI_MODE_TYPE_WATCH))
                || PersistentStore.getBoolean(CollectionServiceStarter.pref_run_wear_collector));
        if (d) Log.d(TAG, "shouldServiceRun() returning: " + result);
        return result;
    }`

Will the previous trick from **Configuration.UI_MODE_TYPE_WATCH;** to **getPackageManager().hasSystemFeature(PackageManager.FEATURE_WATCH);** help in this case?

**Reply by tolot27** [reply]
Yes, it could work. Please replace the function with:

```java
    @SuppressLint("ObsoleteSdkInt")
    private boolean shouldServiceRun() {
        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) return false;
        final boolean result = (DexCollectionType.hasXbridgeWixel() || DexCollectionType.hasBtWixel())
                && ((!Home.get_forced_wear() && !getPackageManager().hasSystemFeature(PackageManager.FEATURE_WATCH))
                || PersistentStore.getBoolean(CollectionServiceStarter.pref_run_wear_collector));
        if (d) Log.d(TAG, "shouldServiceRun() returning: " + result);
        return result;
    }
```

It's important to remove the static declaration. Otherwise, it won't compile.

**Reply by parapenT1sta** [reply]
@tolot27 You just made day! All the issues are fixed with this smartwatch :) It's not even mine but I was trying to help the person to have it working and with you help I was able to! Thank you very much.
I will keep my own version on this watch unless you are thinking to merge your code in to the official release.

Thanks a lot, once again!

**Reply by tolot27** [reply]
Cool, great! Many thanks for investigating this problem and tracking it down to the relevant code lines. That helped quite a lot.

Before I can create a PR, someone has to test it on other devices. Also, we should apply the changes to the wear/src as well and test whether it works on it as well.

**Reply by parapenT1sta** [reply]
@tolot27 I was just following the message from event logger and then verifying the issue was the Bluetooth Service was forcing to shutdown. 
How can we make it available to more people? I can share the APK file here or do you want me to share it on xDrip Facebook Group and ask to have more people testing it?

**Reply by tolot27** [reply]
Let me adapt the changes and create an APK.

**Reply by parapenT1sta** [reply]
Thank you ! :-) I'll be waiting :-)

**Reply by parapenT1sta** [reply]
I've testing and has been working fine.

I've search for Configuration.UI_MODE_TYPE_WATCH and it's also used in JoH but I think is not necessary to change there. [Configuration.UI_MODE_TYPE_WATCH](https://github.com/NightscoutFoundation/xDrip/search?q=Configuration.UI_MODE_TYPE_WATCH)

![PXL_20211228_131011507](https://user-images.githubusercontent.com/608650/147576075-dfc95de1-56aa-4c06-ba3a-1dcc22ddbf81.jpg)

**Reply by parapenT1sta** [reply]
@tolot27 As per this comment, this code is also working for **rogbid air android 9.1 smartwatch**
![image](https://user-images.githubusercontent.com/608650/151038761-9e0baef4-45c5-4747-8757-11748c909c06.png)

**Comment by tolot27** [comment]
https://github.com/NightscoutFoundation/xDrip/pull/1988 was merged today.

**Reply by parapenT1sta** [reply]
Thank youuu!!!!
